### PR TITLE
[RFC] reactor: optionally print decoded stack traces on abort

### PR DESCRIPTION
**Pull request by michoecho**
Decoding stack traces requires access to all binaries appearing in the backtrace, and a manual invocation of seastar-addr2line on the backtrace cop-pasted from logs. This is very inconvenient, especially since seastar became a shared library in debug mode, which makes the main executable alone insufficient to decode a backtrace.

In Scylla, we have backtrace.scylladb.com to automate this, but it is only applicable to public builds, not to local or CI builds.

This series adds an option, implemented using libbacktrace, which prints a decoded backtrace in addition to the regular backtrace.
It is only used on aborts, because the decoding can cause huge stalls.

### CODE REVIEW

**Comment by michoecho** [comment]
Marked as draft because I think it's slightly too hacky at the moment, but the PR is open to comments.

(And the hacky part is that it should be checking whether libbacktrace features are supported before calling it, but that requires including a header generated by `./configure`, and I have no idea how to encode that dependency in cmake. In general, I have no idea what I'm doing with the cmake part of the PR.)

**Comment by michoecho** [comment]
@avikivity @tchaikov Take a look. This is a convenience patch to make dealing with failing tests slightly more convenient (and to restore the ability to decode `debug` mode backtraces broken by the libseastar.so transition).

Don't mind the technicalities yet, just say if you think the idea is reasonable.

File: .gitmodules [hunk 1]
```diff
@@ -1,3 +1,6 @@
 [submodule "dpdk"]
 	path = dpdk
 	url = ../dpdk
+[submodule "libbacktrace"]
+	path = libbacktrace
+	url = https://github.com/ianlancetaylor/libbacktrace
```
Comment (avikivity) [inline .gitmodules:6]: Must we? Is there no packaged alterantive?

Note we're getting std::stacktrace() soon. https://en.cppreference.com/w/cpp/utility/basic_stacktrace

**Comment by michoecho** [comment]
> Usually local machines don't have debug information installed.

That's why it's mainly meant for testing.
It's not often applicable, especially since we build most tests without debug info too, but when applicable, it removes a step that you have to do manually. And I don't think there are any downsides.

File: .gitmodules (hunk 1 above)
Comment (michoecho) [reply .gitmodules:6]: I looked, but I didn't find any packages.

> soon

Say that again after C++20 has a fully-supporting compiler.

**Comment by avikivity** [comment]
But isn't most development done in dev mode?

However, we may be able to restore debug information to dev mode. If we build the application as a shared library, then the debug symbols don't need to be copied around as much.

I don't object to the feature, but I want to be sure it's useful.

File: .gitmodules (hunk 1 above)
Comment (avikivity) [reply .gitmodules:6]:
```c++
$ [ -e /usr/include/c++/12/stacktrace ] && echo yay
yay
```

Comment (michoecho) [reply .gitmodules:6]: > /usr/include/c++/12/stacktrace

Unsurprisingly, it's a wrapper

**Comment by michoecho** [comment]
> But isn't most development done in dev mode?

Is it? Why would you do development in a build mode you can't debug?

**Comment by denesb** [comment]
> > But isn't most development done in dev mode?
> 
> Is it? Why would you do development in a build mode you can't debug?

Because it is faster to build. I do however often add `-g` to `--cflags` these days so I can even debug if I have to.

File: .gitmodules (hunk 1 above)
Comment (tchaikov) [reply .gitmodules:6]: > ```c++
> $ [ -e /usr/include/c++/12/stacktrace ] && echo yay
> yay
> ```

`std::stacktrace` is only available in C++23, quote from `/usr/include/c++12/stacktrace`:

> ```c++
> #if __cplusplus > 202002L && _GLIBCXX_HAVE_STACKTRACE
> ```

so, we are allowed to use C++23 features in Seastar now? or, IOW, are we allowed to use it when the tree is compiled with `-std=c++2b` ? please note, libc++ does not have it at this moment, as per https://en.cppreference.com/w/cpp/compiler_support/23 .

File: CMakeLists.txt [hunk 2]
```diff
@@ -397,6 +397,7 @@ find_package (ragel 6.10 REQUIRED)
 find_package (Threads REQUIRED)
 find_package (PthreadSetName REQUIRED)
 find_package (Valgrind REQUIRED)
+find_package (libbacktrace REQUIRED)
```
Comment (tchaikov) [inline CMakeLists.txt:400]: as it depends on a third party library which is normally not packaged by the distro. IMHO, this feature / dependency should be optional. it should follow the model of how dpdk is supported.

File: cmake/Findlibbacktrace.cmake [hunk 3]
```diff
@@ -0,0 +1,19 @@
+include(GNUInstallDirs)
```
Comment (tchaikov) [inline cmake/Findlibbacktrace.cmake:1]: we should prepare this dependency using the recipe machinery. `Findfoo.cmake` is supposed to *find* a library not for building it.

File: include/seastar/util/print_safe.hh [hunk 4]
```diff
@@ -84,14 +84,14 @@ char* convert_hex_safe(char *buf, size_t bufsz, Integral n) noexcept {
 template<typename Integral>
 SEASTAR_CONCEPT( requires std::integral<Integral> )
 void convert_zero_padded_hex_safe(char *buf, size_t bufsz, Integral n) noexcept {
-    convert_hex_safe<'0'>(buf, bufsz, n);
+    convert_hex_safe<Integral, '0'>(buf, bufsz, n);
```
Comment (tchaikov) [inline include/seastar/util/print_safe.hh:87]: why? i think the compiler should be able to figure out the first template paremeter for `convert_hex_safe<>`, right?

File: include/seastar/util/print_safe.hh
Comment (tchaikov) [inline include/seastar/util/print_safe.hh:1]: could you extract this change into a separate PR?

File: src/util/libbacktrace.h [hunk 5]
```diff
@@ -0,0 +1 @@
+../../libbacktrace/backtrace.h
```
Comment (tchaikov) [inline src/util/libbacktrace.h:1]: i think a better way is to have a proper `Findlibbacktrace.cmake` which exports `libbacktrace::libbacktrace`, and link `seasetar` against it if `Seastar_BACKTRACE` is set. yeah, we need a dedicated option for this feature.

File: src/util/backtrace.cc [hunk 6]
```diff
@@ -30,7 +30,8 @@
 #include <seastar/core/print.hh>
 #include <seastar/core/thread.hh>
 #include <seastar/core/reactor.hh>
-
+#include "libbacktrace.h"
```
Comment (tchaikov) [inline src/util/backtrace.cc:33]: might need to use angle bracket, like

```c++
#include <libbacktrace.h>
```

and guard it using a macro.

File: include/seastar/util/print_safe.hh (hunk 4 above)
Comment (michoecho) [reply include/seastar/util/print_safe.hh:87]: No. How is the compiler supposed to know that `0` is the *second* template argument?

Only trailing arguments can be deduced.

**Comment by michoecho** [comment]
Abandoning due to lack of interest.

File: include/seastar/util/print_safe.hh
Comment (michoecho) [reply include/seastar/util/print_safe.hh:1]: Oh, sorry. I didn't notice this request at the time.
